update gCP levels and extend D4 defaults

hokru opened the pull request:
<!-- Thank you for your contribution! -->

## Description
For psi4 to run B97-3c and r2SCAN-3c the gCP harness needs an update to the allowed methods (limited to the new 'mctc-gcp'). (https://github.com/psi4/psi4/pull/2842) r2SCAN-3c requires manipulation of additional parameters. `dftd4` has been updated to allow this. The D4 harness update sets the additional defaults. Because the dftd4 toml file does not contain these parameters the D4 harness sets them.

Closes #392 

## Changelog description
<!-- Provide a brief single sentence for the changelog. -->
Adds `b973c` and `r2scan3c` methods to the gcp and dftd4 harness.

## Status
<!-- Please `pip install .[lint]; make format` in the base folder. -->
- [x] Code base linted
- [ ] Ready to go

loriab commented:
Ok, this will need a rebase. And possibly you'll want to add a dftd4_350 to testing.py

hokru commented:
From my tests this PR would not work with dftd4 < v.3.5.0. To keep backwards compatibility I will add a check of the dftd4 version and set D4 "default" parameters accordingly.

hokru commented:
Rebased and added a `dftd4_350`. Also made the D4 harness compatible with older dftd4 versions. Can't seem to import `testing._programs` so added a manual solution to check the version.

awvwgk commented on qcengine/testing.py:
```diff
@@ -157,6 +157,7 @@ def get_job(self):
     "dftd3": which("dftd3", return_bool=True),
     "dftd3_321": is_program_new_enough("dftd3", "3.2.1"),
     "dftd4": which_import("dftd4", return_bool=True),
+    "dftd4_350": is_program_new_enough("dftd", "3.5.0"),
```
Looks like a typo?

```suggestion
    "dftd4_350": is_program_new_enough("dftd4", "3.5.0"),
```

awvwgk commented on qcengine/programs/empirical_dispersion_resources.py:
```diff
@@ -891,6 +902,21 @@
+    from dftd4 import __version__ as d4_version
+
+    new_d4_api = parse_version(d4_version) >= parse_version("3.5.0")
+except ModuleNotFoundError:
+    new_d4_api = False
+except ImportError:
+    # handles when no dftd4 present
+    new_d4_api = False
```
```suggestion
except (ModuleNotFoundError, ImportError):
    new_d4_api = False
```

awvwgk commented on qcengine/programs/empirical_dispersion_resources.py:
```diff
@@ -891,6 +902,21 @@
+    new_d4_api = False
+
+# different defaults for dftd4 versions < 3.5.0
+if new_d4_api is False:
```
```suggestion
if not new_d4_api:
```

awvwgk commented on qcengine/programs/gcp.py:
```diff
@@ -160,16 +160,25 @@ def build_input(
             # 'file',
         ]
+        # some methods not available in legacy version
+        mctc_gcp_levels = ["B973C", "R2SCAN3C"]
```
B97-3c should be supported with the original version of gcp already.

loriab replied:
Holger was pretty sure otherwise. See "Dev notes" at https://github.com/psi4/psi4/pull/2842#issue-1518829460 . v2.0.2 at https://anaconda.org/psi4/gcp/files is the relevant "classic" one that I'm trying to wean psi4 folks off of. mctc-gcp is preferred as of psi4 v1.8.

awvwgk commented on qcengine/programs/empirical_dispersion_resources.py:
```diff
@@ -5,6 +5,7 @@
 from typing import Dict, List, Optional, Union
 
 from ..exceptions import InputError
+from pkg_resources import parse_version
```
Try `qcelemental.util.parse_version` instead.

loriab commented:
Thanks, @awvwgk, I think I've implemented all your suggestions here, except the b973c uncertainty (which I don't mind using as an extra push for users to switch).

I think this is ready to go, unless someone posts reservations. My plan is to merge it tomorrow after I get a clean CI at https://github.com/psi4/psi4/pull/2842 .
